File: 3350/adjacent-increasing-subarrays-detection-ii.py
# LeetCode 3350. Adjacent Increasing Subarrays Detection II
# 這題跟 3349 Easy 題非常相似，但差在 k 的長度「要你自己找出來」
# 最多有 20萬個數「不能用暴力for迴圈」。可用「左到右combo」「右到左combo」
class Solution:
    def maxIncreasingSubarrays(self, nums: List[int]) -> int:
        N = len(nums)  # 陣列的長度
        left, right = [1]*N, [1]*N  # 「左到右combo」「右到左combo」
        for i in range(1,N):  # 建立「左到右combo」
            if nums[i-1]<nums[i]: left[i] = left[i-1] + 1
        for i in range(N-1-1, -1, -1):  # 反過來，建立「右到左combo」
            if nums[i]<nums[i+1]: right[i] = right[i+1] + 1
        ans = 0
        for i in range(N-1):  # 迴圈巡一輪，找到最大的 k 值
            now = min(left[i], right[i+1])  # 現在這格的 k 值
            ans = max(ans, now)  # 更新答案
        return ans   
        

File: 3350/adjacent-increasing-subarrays-detection-ii.cpp
// LeetCode 3350. Adjacent Increasing Subarrays Detection II
// 題目跟昨天的 3349 很像。昨天給你 k 今天問你「最大可能的k」
// 3349用 combo, 今天 3350 用兩個combo 左到右、右到左,累積多少
class Solution {
public:
    int maxIncreasingSubarrays(vector<int>& nums) {
        int N = nums.size(); // 陣列的大小
        vector<int> left(N, 1); // 左到右的combo
        vector<int> right(N, 1); // 右到左的combo
        for(int i=1; i<N; i++) { // 左到右的 combo 逐一更新
            if( nums[i-1] < nums[i] ) left[i] = left[i-1] + 1;
        } // 如果有連續, 就更新、變更大
        for(int i=N-2; i>=0; i--) { // 右到左的combo 逐一更新
            if( nums[i] < nums[i+1]) right[i] = right[i+1] + 1;
        }
        int ans = 0; // 要找到最大的k,也就是「兩個相鄰陣列」要持續增加的最長的長度k
        for(int i=0; i<N-1; i++) {
            int now = min(left[i], right[i+1]); // 現在這格左邊combo 右邊combo 都符合
            if(now>ans) ans = now; // 有更大的k,就用來更新ans
        }
        return ans;
    }
};
